srcjs/shinyapp.js

```diff
@@ -631,6 +631,38 @@ var ShinyApp = function() {
     window.location.reload();
   });
 
+  addMessageHandler('shiny-insert-ui', function (message) {
+      var targets = $(message.selector);
+      if (targets.length === 0) {
+        // render the HTML and deps to a null target, so
+        // the side-effect of rendering the deps, singletons,
+        // and <head> still occur
+        exports.renderHtml($([]), message.content.html, message.content.deps);
+      } else {
+        targets.each(function (i, target) {
+          var container = document.createElement(message.container);
+          insertAdjacentElement(message.where, target, container);
+          exports.renderContent(container, message.content);
+          $(container).trigger('shown');
+          return message.multiple;
+        });
+      }
+    });
+
+  addMessageHandler('shiny-remove-ui', function (message) {
+    var els = $(message.selector);
+    els.each(function (i, el) {
+      exports.unbindAll(el, true);
+      $(el).trigger('hide');
```

Comment:
I think this `hide` and `hidden` stuff isn't necessary, for a couple reasons. From a conversation, I understand that the reason this is here is so that `sendImageSize` and `sendOutputHiddenState` get triggered when a UI element is removed. There are two reasons I don't think that's needed:

1. Those functions update the clientData width and height, and hidden state. Those values are only needed for re-rendering a plot, but if the plot output UI is removed, there's no need to update those values.
2. In the `htmlOutputBinding` (`uiOutput` + `renderUI`), it's also possible to remove a plot output, by simply replacing it with nothing. But in that case, we don't bother triggering `hide` or `hidden` events, and everything seems to work fine.